crates/win-api-wrappers/src/rpc.rs

```diff
@@ -0,0 +1,214 @@
+            Rpc::{
+                RpcBindingVectorFree, RpcImpersonateClient, RpcRevertToSelfEx, RpcServerInqBindings,
+                RpcServerInqCallAttributesW, MIDL_SERVER_INFO, RPC_BINDING_VECTOR, RPC_CALL_ATTRIBUTES_V2_W,
+                RPC_CALL_ATTRIBUTES_VERSION, RPC_QUERY_CLIENT_PID, RPC_QUERY_CLIENT_PRINCIPAL_NAME,
+                RPC_QUERY_SERVER_PRINCIPAL_NAME, RPC_SERVER_INTERFACE, SERVER_ROUTINE,
+            },
+        },
+    },
+};
+
+use crate::win::set_memory_protection;
+use crate::{
+    error::Error,
+    win::{Thread, Token},
+};
+
+use anyhow::{bail, Result};
+
+#[allow(non_camel_case_types)]
+pub type RPC_BINDING_HANDLE = *mut c_void;
+
+pub struct RpcBindingHandle(pub RPC_BINDING_HANDLE);
+
+pub struct RpcCallAttributes {
+    pub client_pid: u32,
+    pub server_principal_name: String,
+    pub client_principal_name: String,
+}
+
+impl RpcBindingHandle {
+    pub fn inquire_caller(&self) -> Result<RpcCallAttributes> {
+        let mut client_principal_name = Vec::new();
+        let mut server_principal_name = Vec::new();
+
+        let mut attribs = RPC_CALL_ATTRIBUTES_V2_W::default();
+        attribs.Version = RPC_CALL_ATTRIBUTES_VERSION;
+        attribs.Flags = RPC_QUERY_SERVER_PRINCIPAL_NAME | RPC_QUERY_CLIENT_PRINCIPAL_NAME | RPC_QUERY_CLIENT_PID;
+        let status = unsafe { RpcServerInqCallAttributesW(Some(self.0), &mut attribs as *mut _ as _) };
+
+        if status.0 != ERROR_MORE_DATA.0 as i32 {
+            bail!(Error::from(status));
+        }
+
+        client_principal_name.resize((attribs.ClientPrincipalNameBufferLength / 2) as _, 0);
+        server_principal_name.resize((attribs.ServerPrincipalNameBufferLength / 2) as _, 0);
+
+        attribs.ClientPrincipalName = client_principal_name.as_mut_ptr();
+        attribs.ClientPrincipalNameBufferLength = (client_principal_name.len() * mem::size_of::<u16>()) as _;
+        attribs.ServerPrincipalName = server_principal_name.as_mut_ptr();
+        attribs.ServerPrincipalNameBufferLength = (server_principal_name.len() * mem::size_of::<u16>()) as _;
+
+        let status = unsafe { RpcServerInqCallAttributesW(Some(self.0), &mut attribs as *mut _ as _) };
+
+        status.ok()?;
+
+        // Remove trailing null byte
+        client_principal_name.pop();
+        server_principal_name.pop();
+
+        Ok(RpcCallAttributes {
+            client_pid: attribs.ClientPID.0 as _,
+            client_principal_name: String::from_utf16(&client_principal_name)
+                .map_err(|e| windows::core::Error::from(e))?,
+            server_principal_name: String::from_utf16(&server_principal_name)
+                .map_err(|e| windows::core::Error::from(e))?,
+        })
+    }
+
+    pub fn impersonate_client<F>(&self, f: F) -> Result<()>
+    where
+        F: FnOnce() -> Result<()>,
+    {
+        unsafe { RpcImpersonateClient(Some(self.0)) }.ok()?;
+
+        let r = f();
+
+        unsafe { RpcRevertToSelfEx(Some(self.0)) }.ok()?;
```

Comment:
It would be great to wrap this  in some kind of wrapper, which calls RpcRevertToSelfEx on destructor